### ISSUE TITLE
Accessibility of forms

### ISSUE THREAD

**Issue by mcrossley** [jsonform/jsonform#341]
I am looking for a package just like this to replace Alpaca forms. I have prototyped a simple page to test this package out...

But like Alpaca it seems to have issues with accessibility. Collapsed field sets for instance do not seem to be expandable using the keyboard, only a click event is used?

Is there a way of enabling additional navigation options?

I have some blind users that have to use screen readers and keyboard navigation, and my application has some quite complex settings screens using JSON based forms. I also use collapsed and conditionally visible field sets quite extensively.

**Comment by sdetweil** [comment]
are they using jaws or built in navigation? 

i thought I might have add something, but not I don't think.. 
where are classes expandable and expanded (added when expanded) 

but the accessibility spec says you would need a button there to be a focus target, and operation target..

I thought prepend might work, but its only for input fields. and the expandable heading, fieldset, is not an input field.. 
course a tiny mod to the jsonform library could make it otherwise.. 

i made one to allow disabling drag on tabarry..

**Comment by sdetweil** [comment]
i was able to change the legend to enclose a button, which make the expandable sections keyboard navigable.
i added a new option and updated the fieldset template

i could add this to my JSONFORM fork, and u could try it..  i only did fieldset not the other two auth and select
but those would be easy, given what I did.

there is a visual change as the text is now the button text 

![Screenshot at 2021-05-05 15-22-11](https://user-images.githubusercontent.com/1498142/117204277-b8708600-adb5-11eb-932b-89df54ae6e3f.png)

**Comment by sdetweil** [comment]
I pushed this change to my fork 
https://github.com/sdetweil/jsonform

set 
accessible:true 
on the fieldset definition 
here is my template for those headings
```
const module_form_template= {
							             "type": "fieldset",
							             "title": "modulename",
							             //"htmlClass":"module_name_class",
							             "expandable": true,
								    "accessible": false,   // <--- new parm, set to true for button for legend
   								     "items": []
}
```

**Comment by mcrossley** [comment]
Great, I'll take a look at what you have done - it will be tomorrow now.

I was using MS Narrator, and no navigation was possible. But I then tried NVDA, and when that is running the fieldsets are expandable using <space> or <return>.

I have also been playing around with adding Aria attributes.
For instance on the <legend>, make it <legend role="treeitem" aria-expanded="false">...
And NVD reads it as a clickable, and says it is collapsed. It would then need the jsonform JS to toggle the aria-expanded attribute value to true when the fieldset is expanded.

TBH all these accessibility issues are quite new to me and I'm feeling my way around...

**Comment by sdetweil** [comment]
no problem.. I spent 5 years in the Accessibility group in IBM so I understood right away .. i went and got the jaws inspector to test with.. 

there is a prepend attribute already that lets you inject 'any' html, but it doesn't apply to fieldset, I tried that first, 
but the code would be IN the legend element and can't wrap it.. if u add stuff..

**Comment by sdetweil** [comment]
also there is an existing class added on expand, 
'expanded'

so you could override that and add css elements like the aria attributes

the docs show examples of using expanded class override to make the twisty thing in front..

**Comment by mcrossley** [comment]
OK, I took a look at your updates and merged in my aria attributes. It works as far as display goes, but the fieldset buttons are also triggering the onSubmit function :(

BTW, I think the doc is slightly wrong for the fieldset style suggestions, they should be...
`.expandable > legend::before {
    content: '\25B8';
    padding-right: 5px;
}
.expanded > legend::before {
    content: '\25BE';
}
`
Note the double colons rather than single.

**Comment by mcrossley** [comment]
OK, I think I have it sorted (for fieldsets anyway)...

[Accessibility enabled in JSON data](https://weather.wilmslowastro.com/test/mxsettings/stationsettings.htm)
[Accessibility disabled in JSON data](https://weather.wilmslowastro.com/test/mxsettings/stationsettings2.htm)

The jsonform.js file is your new base with my aria "tweaks".

What I would like to try and sort out next is dynamically rebuilding the form if the accessibility checkbox is changed without having to submit the values and reload the page...

Then move on to some of my more advanced pages and find any other accessibility issues - after I figure out how to build them!

**Comment by sdetweil** [comment]
as for the expandable info from the doc, copy/pasted it into my form css as is and it did what I wanted. gave the twisty out front in two different shapes.

**Comment by sdetweil** [comment]
I don't get submit with my button added, just expand or not.

**Comment by sdetweil** [comment]
I don't understand the 
accessible: {{values.accessible}}

we can talk offline. my email is my same user ID at gmail

**Comment by tchapi** [comment]
Hi @sdetweil  and @mcrossley 

Thanks for this issue and the discussion around it
This accessibility thing is really interesting and I think it would be great if it could be supported directly in jsonform.

Is including a button in the legend the only way to make the fieldset accessible-compliant to use with screen readers and keyboards ? It feels a bit hacky — whereas adding `aria-*`  attributes like proposed (`legend role="treeitem" aria-expanded="false">`) seems more in line with what the lib should do.

I could spend some time on this to include it if you don't mind telling me your feelings about the changes and how to best make it compliant

looking forward;

**Comment by mcrossley** [comment]
TBH, at this stage I don't know. I've gone with a combination of the button and aria attributes for now - [latest incarnation supports dynamic switching](https://weather.wilmslowastro.com/test/mxsettings/stationsettings3.htm).

I think it depends on the reader used. I tested NVDA for instance, and with that the original jsonform code was usable (just) without any changes, adding the aria attributes makes it much clearer what is happening. NVDA allows you to use the Space or Return to expand the fieldset. But if you tried MS Narrator it just says there is no clickable element on the collapsed legends and you are stuck.

**Comment by tchapi** [comment]
Ok — seems that adding the `aria-*` attributes is a quick win anyway, no ?

**Comment by tchapi** [comment]
I've pushed a PR to add the `aria-expanded` attributes — hope it helps !

**Comment by tchapi** [comment]
Closing for now — tell me if this needs reopening
